reconcile/dashdotdb_dora.py

```diff
@@ -0,0 +1,503 @@
+import functools
+import os
+from collections import defaultdict
+from dataclasses import dataclass
+from datetime import (
+    datetime,
+    timedelta,
+    timezone,
+)
+from typing import (
+    Any,
+    Optional,
+)
+
+import requests
+import yaml
+from psycopg2 import (
+    connect,
+    sql,
+)
+from sretoolbox.utils import threaded
+
+from reconcile import queries
+from reconcile.dashdotdb_base import (
+    LOG,
+    DashdotdbBase,
+)
+from reconcile.typed_queries.app_interface_vault_settings import (
+    get_app_interface_vault_settings,
+)
+from reconcile.typed_queries.saas_files import get_saas_files
+from reconcile.utils.github_api import GithubRepositoryApi
+from reconcile.utils.gitlab_api import GitLabApi
+from reconcile.utils.secret_reader import create_secret_reader
+
+
+@dataclass
+class DeploymentDBSecret:
+    path: str
+    field: str
+    q_format: Optional[str]
+    version: Optional[int]
+
+
+DEPLOYMENT_DB_SECRET = DeploymentDBSecret(
+    field="",
+    path=os.environ["DEPLOYMENT_DB_SECRET"],
+    q_format=None,
+    version=None,
+)
+
+QONTRACT_INTEGRATION = "dashdotdb-dora"
+
+
+@dataclass(eq=True, frozen=True)
+class AppEnv:
+    app_name: str
+    env_name: str
+
+
+@dataclass
+class Deployment:
+    trigger_reason: str
+    deployment_time: datetime
+
+
+class DeploymentDB:
+    # vault: app-sre/integrations-output/terraform-resources/appsrep05ue1/app-sre-pipeline-utils/deploy-events-db
+    def __init__(
+        self,
+        host: str,
+        port: str,
+        name: str,
+        user: str,
+        password: str,
+    ):
+        self.conn = connect(
+            "host={} port={} dbname={} user={} password={}".format(
+                host, port, name, user, password
+            ),
+        )
+        LOG.info("Connected to DeploymentDB")
+
+    def deployments(
+        self,
+        trigger_reason_pattern: str,
+        app_env_since_list: list[tuple[AppEnv, datetime]],
+    ) -> list[tuple[AppEnv, Deployment]]:
+        deployments = []
+        with self.conn.cursor() as cur:
+            subq = [
+                sql.SQL(
+                    "(app_name = {} AND env_name = {} AND deployment_time > {})"
+                ).format(
+                    sql.Literal(app_env.app_name),
+                    sql.Literal(app_env.env_name),
+                    sql.Literal(since),
+                )
+                for (app_env, since) in app_env_since_list
+            ]
+
+            query = sql.SQL(
+                """
+                    SELECT app_name, env_name, trigger_reason, deployment_time
+                    FROM deployments
+                    WHERE
+                        succeeded = True
+                        AND trigger_reason LIKE {}
+                        AND ({})
+                    ORDER BY deployment_time ASC;
+                """
+            ).format(sql.Literal(trigger_reason_pattern), sql.SQL(" OR ").join(subq))
```

Comment:
This query is a little complex, and I'm not sure if this can work with large dataset as this query has too many conditions, hard to use index.

I'm feeling we are missing a domain concept here, the logic looks like 

```
Return all app-sre managed unregistered deployment info
```

I'm not sure if deployments table contains any non app-sre managed deployment info, if none, then we don't need that filter. If yes, then maybe a new column `managed_by` value `app-sre` can simplify this query.

There is another filter about `deployment_time`, if we can introduce another column `registered_time`, populate it after register in dashdotdb, then the whole query will be like this:


```python
                    SELECT app_name, env_name, trigger_reason, deployment_time
                    FROM deployments
                    WHERE
                        succeeded = True
                        AND trigger_reason LIKE {}
                        AND managed_by = {}
                        AND registered_time is NULL
                    ORDER BY deployment_time ASC;
```

Index on `managed_by, registered_time` will speed up this query.

For first time migration, if we want to skip old deployments, can populate `registered_time` to a value or not set `managed_by`.

This is based on we control deployments database scheme, if it's out of our control, then the complex query is needed.

Reply:
This is probably premature optimization. The query takes `0:00:00.592230` on the first run, and `0:00:00.252267` in subsequent ones, so I think this is reasonable for now. If it's OK with you I'd prefer not to solve this problem at this stage.

Reply:
250ms is a very slow query in db, if this feature is not designed to scale, then I'm ok with this.

Reply:
Thanks. I agree 250ms is not ideal, but the runtime here is not important. We are going to run this once per day.